In [50]:
import numpy as np
import pandas as pd

In [51]:
data = pd.read_csv('train.csv')

In [52]:
data.head(n=10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [53]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [54]:
columns_to_drop = ['PassengerId','Name','Ticket','Cabin','Embarked']

In [55]:
data_clean = data.drop(columns_to_drop,axis=1)

In [56]:
data_clean.head(n=5)

Survived  Pclass     Sex   Age  SibSp  Parch     Fare
0         0       3    male  22.0      1      0   7.2500
1         1       1  female  38.0      1      0  71.2833
2         1       3  female  26.0      0      0   7.9250
3         1       1  female  35.0      1      0  53.1000
4         0       3    male  35.0      0      0   8.0500

In [57]:
data_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    object 
 3   Age       714 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
dtypes: float64(2), int64(4), object(1)
memory usage: 48.9+ KB


In [58]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
data_clean['Sex'] = le.fit_transform(data_clean['Sex'])

In [59]:
data_clean.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare
0         0       3    1  22.0      1      0   7.2500
1         1       1    0  38.0      1      0  71.2833
2         1       3    0  26.0      0      0   7.9250
3         1       1    0  35.0      1      0  53.1000
4         0       3    1  35.0      0      0   8.0500

In [60]:
data_clean.info()
# notice all data are either int or float

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    int32  
 3   Age       714 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
dtypes: float64(2), int32(1), int64(4)
memory usage: 45.4 KB


In [61]:
# Filling missing age values by mean
data_clean = data_clean.fillna(data_clean["Age"].mean())

In [62]:
data_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    int32  
 3   Age       891 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
dtypes: float64(2), int32(1), int64(4)
memory usage: 45.4 KB


## Now all columns have equal values = 891

### We can also use Imputer from sklearn for this mean() thing

In [63]:
data_clean.shape

(891, 7)

In [64]:
#Getting data from 2nd row
data_clean.loc[2]

Survived     1.000
Pclass       3.000
Sex          0.000
Age         26.000
SibSp        0.000
Parch        0.000
Fare         7.925
Name: 2, dtype: float64

In [65]:
#Getting data from 2nd row and 1st col
data_clean.loc[2][1]

3.0

## Splitting Table into X and Y

In [66]:
input_cols = ['Pclass','Sex','Age','SibSp','Parch','Fare']
output_cols = ['Survived']

X = data_clean[input_cols]
Y = data_clean[output_cols]

print(X.shape,Y.shape)
print(type(X))

(891, 6) (891, 1)
<class 'pandas.core.frame.DataFrame'>


## Define Entropy and Information Gain

In [67]:
def entropy(col):
    
    counts = np.unique(col,return_counts=True)
    N = float(col.shape[0])
    
    ent = 0.0
    
    for ix in counts[1]:
        p = ix/N
        ent += (p*np.log2(p))
        
    return -1.0 * ent

In [68]:
col = np.array([1,1,1,0,0,0])
entropy(col)

1.0

In [69]:
def divide_data(x_data,fkey,fval):
    
    x_right = pd.DataFrame([],columns=x_data.columns)
    x_left = pd.DataFrame([],columns=x_data.columns)
    
    for ix in range(x_data.shape[0]):
        val = x_data[fkey].loc[ix]
        
        if val > fval:
            # Appeding entire Row
            x_right = x_right.append(x_data.loc[ix])
        else:
            x_left = x_left.append(x_data.loc[ix])
            
    return x_left,x_right

In [70]:
x_left,x_right = divide_data(data_clean[:10],'Sex',0.5)
print(x_left)
print(x_right)

   Survived  Pclass  Sex   Age  SibSp  Parch     Fare
1       1.0     1.0  0.0  38.0    1.0    0.0  71.2833
2       1.0     3.0  0.0  26.0    0.0    0.0   7.9250
3       1.0     1.0  0.0  35.0    1.0    0.0  53.1000
8       1.0     3.0  0.0  27.0    0.0    2.0  11.1333
9       1.0     2.0  0.0  14.0    1.0    0.0  30.0708
   Survived  Pclass  Sex        Age  SibSp  Parch     Fare
0       0.0     3.0  1.0  22.000000    1.0    0.0   7.2500
4       0.0     3.0  1.0  35.000000    0.0    0.0   8.0500
5       0.0     3.0  1.0  29.699118    0.0    0.0   8.4583
6       0.0     1.0  1.0  54.000000    0.0    0.0  51.8625
7       0.0     3.0  1.0   2.000000    3.0    1.0  21.0750


In [71]:
def information_gain(x_data,fkey,fval):
    
    left,right = divide_data(x_data,fkey,fval)
    
    # %ge of total samples on left and right
    l = float(left.shape[0])/x_data.shape[0]
    r = float(right.shape[0])/x_data.shape[0]
    
    # All examples come to one side
    if left.shape[0] == 0 or right.shape[0] == 0:
        return -1000000 # Min Info Gain
    
    i_gain = entropy(x_data.Survived) - (l*entropy(left.Survived) + r*entropy(right.Survived)) 
    
    return i_gain

In [72]:
# Test our Function
for fx in X.columns:
    print(fx,information_gain(data_clean,fx,data_clean[fx].mean()))

Pclass 0.07579362743608165
Sex 0.2176601066606142
Age 0.0008836151229467681
SibSp 0.009584541813400071
Parch 0.015380754493137694
Fare 0.042140692838995464


## Implementing Decision Tree

In [73]:
class DecisionTree:
    
    # constructor
    def __init__(self,depth=0,max_depth=5,left="Nan",right="Nan"):
        self.left = None
        self.right = None
        self.fkey = None
        self.fval = None
        self.max_depth = max_depth
        self.depth = depth
        self.terget = None
        self.bool_left = left
        self.bool_right = right
        
    def train(self,x_train):
        
        features = ['Pclass','Sex','Age','SibSp','Parch','Fare']
        info_gains = []
        
        for ix in features:
            i_gain = information_gain(x_train,ix,x_train[ix].mean())
            info_gains.append(i_gain)
            
        self.fkey = features[np.argmax(info_gains)]
        self.fval = x_train[self.fkey].mean()
        print('Making Tree.. Depth is %d, Feature is %s, left: %s, Right: %s'%(self.depth,self.fkey,self.bool_left, self.bool_right))
        
        # Split data
        data_left, data_right = divide_data(x_train,self.fkey,self.fval)
        data_left = data_left.reset_index(drop=True)
        data_right = data_right.reset_index(drop=True)
        
        
        # Reached Leaf Node
        if data_left.shape[0] == 0 or data_right.shape[0] == 0:
            if x_train.Survived.mean() >= 0.5:
                self.target = "Survive"
            else:
                self.target = "Dead"
            return
        
        # Stop early when depth >= max depth
        if self.depth >= self.max_depth:
            if x_train.Survived.mean() >= 0.5:
                self.target = "Survive"
            else:
                self.target = "Dead"
            return
        
        # Build tree recursively
        self.left = DecisionTree(depth=self.depth+1,max_depth=self.max_depth,left="True",right="False")
        self.left.train(data_left)
        
        self.right = DecisionTree(depth=self.depth+1,max_depth=self.max_depth,left="False",right="True")
        self.right.train(data_right)
        
        # Setting target at every Node
        if x_train.Survived.mean() >= 0.5:
            self.target = "Survive"
        else:
            self.target = "Dead"
        return
    
    
    def predict(self,test):
        
        if test[self.fkey] >= self.fval:
            # Go to right
            if self.right is None:
                return self.target
            return self.right.predict(test)
        else:
            if self.left is None:
                return self.target
            return self.left.predict(test)
            
        
            
        
            

In [74]:
d_tree = DecisionTree()
d_tree.train(data_clean)

Making Tree.. Depth is 0, Feature is Sex, left: Nan, Right: Nan
Making Tree.. Depth is 1, Feature is Pclass, left: True, Right: False
Making Tree.. Depth is 2, Feature is Pclass, left: True, Right: False
Making Tree.. Depth is 3, Feature is Parch, left: True, Right: False
Making Tree.. Depth is 4, Feature is Age, left: True, Right: False
Making Tree.. Depth is 5, Feature is Age, left: True, Right: False
Making Tree.. Depth is 5, Feature is Age, left: False, Right: True
Making Tree.. Depth is 4, Feature is Parch, left: False, Right: True
Making Tree.. Depth is 5, Feature is Age, left: True, Right: False
Making Tree.. Depth is 5, Feature is Fare, left: False, Right: True
Making Tree.. Depth is 3, Feature is Parch, left: False, Right: True
Making Tree.. Depth is 4, Feature is Age, left: True, Right: False
Making Tree.. Depth is 5, Feature is Age, left: True, Right: False
Making Tree.. Depth is 5, Feature is Age, left: False, Right: True
Making Tree.. Depth is 4, Feature is Age, left: Fals

In [75]:
np.unique(data_clean.Pclass)

array([1, 2, 3], dtype=int64)

## Making Prediction

### Train-Test Split

In [76]:
split = int(0.7*data_clean.shape[0])

train_data = data_clean[:split]
test_data = data_clean[split:]

test_data = test_data.reset_index(drop=True)

print(train_data.shape,test_data.shape)

(623, 7) (268, 7)


In [77]:
print(d_tree.fkey)
print(d_tree.fval)
print(d_tree.left.fkey)
print(d_tree.right.fkey)

Sex
0.6475869809203143
Pclass
Fare


In [78]:
ypred = []
for ix in range(test_data.shape[0]):
    ypred.append(d_tree.predict(test_data.loc[ix]))
    
yactual = test_data[output_cols]

In [79]:
le = LabelEncoder()
ypred = le.fit_transform(ypred)

In [80]:
print(ypred.shape,yactual.shape)

(268,) (268, 1)


In [81]:
ypred = np.array(ypred).reshape((-1,1))
print(ypred.shape,yactual.shape)

(268, 1) (268, 1)


In [82]:
acc = np.sum(ypred==np.array(yactual))/ypred.shape[0]

In [83]:
print(acc)

0.8619402985074627
